concurrency/smallTest/MessagingReducedConcurrentHashMapAVOS01.java

```diff
@@ -0,0 +1,102 @@
+package concurrency.smallTest;
+
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * The MINIMUM reduced messaging.  I.e., the other messaging test has a ton of combinations.  This only the 
+ *  minimum necessary to show all coverage
+ * */
+public class MessagingReducedConcurrentHashMapAVOS01 {
+    private ConcurrentHashMap<String, String> myMap = new ConcurrentHashMap<>();
+
+    String key = "lala" + 33;
+    String val = "lala" + 44;
+    
+    public String give() {
+        myMap.remove("key"); // this remove ensures the noRemove() policy does not kick in
+        return myMap.get("lala");
+    }
+    
+    // =========== check and get ===========
+    
+    public void atomViol06() {
+        if ( myMap.containsKey(key)) {
+            String myVal = myMap.get(key);
+            System.out.println(myVal.length()); // ensure accesses null
+        }
+    }
+    
+    // ================ check and remove ===============
+    
+    public void atomViol_checkRemove_01() {
+        if (myMap.get(key) != null) {
+            String myVal = myMap.remove(key);
+            System.out.println(myVal.size());            
+        }
+    }
+    
+    // ========= check and put with additional info about the put =========
+    public void atomViol07() {
+        String myVal = myMap.get(key);
+        if (myVal == null) {
+            myMap.put(key, val);
+        }
+        return val;
+    }
+    
+    public void atomViol08() {
+        if (myMap.get(key) == null) {
+            myMap.put(key, val);
+        }
+        System.out.println(val);
+    }
+    
+    // ================== over sync with put ================
+    
+    public String atomViol09() {
+        String myKey = "lala" + 88;
+        String sillyVal = "lala" + 99;
+        synchronized (this) {
+            if ( ! myMap.containsKey(myKey)) {
+                myMap.put(myKey, sillyVal);
+            }
+        }
+        System.out.println(sillyVal);
+        return sillyVal;
+    }
+    
+    // =================== over sync with remove =============
+    
+    public void overSync11() {
+        String myKey = "lala" + 88;
+        String sillyVal = "lala" + 99;
+        synchronized (this) {
+            boolean contains = myMap.containsKey(myKey);
+            if ( ! contains) {
+                myMap.remove(myKey);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 These lines use `ConcurrentHashMap`&#39;s **sequential** API and are protected by `synchronized` to ensure they execute together: **74 and 76**. This synchronization may reduce performance by limiting parallelism and does not take advantage of `ConcurrentHashMap`&#39;s **concurrent** API.

**Fix**  
 Consider replacing the check and `remove(key)` with `remove(key, value)` and then removing the synchronization. If you do not need the check, consider removing the check and the synchronization and just calling `remove(key)`.

**More info**  
 `remove(key, value)` removes the key only if the current value is equal to the `value` parameter. `remove(key, value)` returns `true` if the call removed the `key` and `value`.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/4d3c580b03475a6ec9323d11e6875c542f8e3f6d/hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/Client.java#L1591-L1592) (external link).